core/src/poh_timing_reporter.rs

```diff
@@ -0,0 +1,151 @@
+        write!(
+            f,
+            "SlotPohTimestamp: start={} end={} full={}",
+            self.start_time, self.end_time, self.full_time
+        )
+    }
+}
+
+impl SlotPohTimestamp {
+    /// Return true if the timing points of all events are received.
+    pub fn is_complete(&self) -> bool {
+        self.start_time != 0 && self.end_time != 0 && self.full_time != 0
+    }
+
+    /// Update with timing point
+    pub fn update(&mut self, timing_point: PohTimingPoint) {
+        match timing_point {
+            PohTimingPoint::PohSlotStart(ts) => self.start_time = ts,
+            PohTimingPoint::PohSlotEnd(ts) => self.end_time = ts,
+            PohTimingPoint::FullSlotReceived(ts) => self.full_time = ts,
+        }
+    }
+
+    /// Return the time difference from slot start to slot shred full
+    pub fn slot_start_to_full_time(&self) -> i64 {
+        (self.full_time as i64).saturating_sub(self.start_time as i64)
+    }
+
+    /// Return the time difference from slot shred full to slot end
+    pub fn slot_full_to_end_time(&self) -> i64 {
+        (self.end_time as i64).saturating_sub(self.full_time as i64)
+    }
+}
+
+/// A PohTimingReporter manages and reports the timing of events for incoming
+/// slots
+#[derive(Default)]
+pub struct PohTimingReporter {
+    /// Storage map of SlotPohTimestamp per slot
+    slot_timestamps: HashMap<Slot, SlotPohTimestamp>,
+}
+
+impl PohTimingReporter {
+    /// Return true if PohTiming is complete for the slot
+    pub fn is_complete(&self, slot: Slot) -> bool {
+        if let Some(slot_timestamp) = self.slot_timestamps.get(&slot) {
+            slot_timestamp.is_complete()
+        } else {
+            false
+        }
+    }
+
+    /// Report PohTiming for a slot
+    pub fn report(&self, slot: Slot, slot_timestamp: &SlotPohTimestamp) {
+        datapoint_info!(
+            "poh_slot_timing",
+            ("slot", slot as i64, i64),
+            ("start_time", slot_timestamp.start_time as i64, i64),
+            ("end_time", slot_timestamp.end_time as i64, i64),
+            ("full_time", slot_timestamp.full_time as i64, i64),
+            (
+                "start_to_full_time_diff",
+                slot_timestamp.slot_start_to_full_time(),
+                i64
+            ),
+            (
+                "full_to_end_time_diff",
+                slot_timestamp.slot_full_to_end_time(),
+                i64
+            ),
+        );
+    }
+
+    /// Process incoming PohTimingPoint from the channel
+    pub fn process(&mut self, slot: Slot, t: PohTimingPoint) {
+        let slot_timestamp = self
+            .slot_timestamps
+            .entry(slot)
+            .or_insert_with(SlotPohTimestamp::default);
+
+        slot_timestamp.update(t);
+
+        if let Some(slot_timestamp) = self.slot_timestamps.get(&slot) {
+            if slot_timestamp.is_complete() {
+                self.report(slot, slot_timestamp);
+                //let _ = self.slot_timestamps.remove(&slot);
```

Comment:
remove this line?

Reply:
But, how is the `self.slot_timestamps` cleaned up?

Reply:
yeah. good catch.

Reply:
Is that acceptable? or I could just store the slot_id only? which will be around 1.7 MB a day